walletdb/interface.go

```diff
@@ -207,23 +208,60 @@ type DB interface {
 	Close() error
 }
 
-// BatchDB is a special version of the main DB interface that allos the caller
-// to specify write transactions that should be combine dtoegether if multiple
+// BatchDB is a special version of the main DB interface that allows the caller
+// to specify write transactions that should be combined toegether if multiple
 // goroutines are calling the Batch method.
 type BatchDB interface {
 	DB
 
 	// Batch is similar to the package-level Update method, but it will
-	// attempt to optismitcally combine the invocation of several
+	// attempt to optimistically combine the invocation of several
 	// transaction functions into a single db write transaction.
 	Batch(func(tx ReadWriteTx) error) error
 }
 
+// ExtendedDB is a special version of the main DB interface that provides their
+// own implementation of the View and Update methods. The walletdb package will
+// use those methods instead of the package global View/Update functions if the
+// given DB is compatible with this extended interface.
+type ExtendedDB interface {
+	DB
+
+	// PrintStats returns all collected stats pretty printed into a string.
+	PrintStats() string
+
+	// View opens a database read transaction and executes the function f
+	// with the transaction passed as a parameter. After f exits, the
+	// transaction is rolled back. If f errors, its error is returned, not a
+	// rollback error (if any occur). The passed reset function is called
+	// before the start of the transaction and can be used to reset
+	// intermediate state. As callers may expect retries of the f closure
+	// (depending on the database backend used), the reset function will be
+	//called before each retry respectively.
+	View(f func(tx ReadTx) error, reset func()) error
+
+	// Update opens a database read/write transaction and executes the
+	// function f with the transaction passed as a parameter. After f exits,
+	// if f did not error, the transaction is committed. Otherwise, if f did
+	// error, the transaction is rolled back. If the rollback fails, the
+	// original error returned by f is still returned. If the commit fails,
+	// the commit error is returned. As callers may expect retries of the f
+	// closure (depending on the database backend used), the reset function
+	// will be called before each retry respectively.
+	Update(f func(tx ReadWriteTx) error, reset func()) error
+}
+
 // View opens a database read transaction and executes the function f with the
 // transaction passed as a parameter.  After f exits, the transaction is rolled
 // back.  If f errors, its error is returned, not a rollback error (if any
 // occur).
 func View(db DB, f func(tx ReadTx) error) error {
+	// Allow external/remote DB driver implementations to overwrite the
+	// default view behavior.
+	if edb, ok := db.(ExtendedDB); ok {
```

Comment:
To be fully safe I think we need some deeper changes, because there are some cases in the wallet where in the `View` and `Update` closures mutate collections inside the closure. We won't retry them since everything is serial and we also use the walletdb rwlocked in LND but the `reset` closure could be used to reset those collections where needed.

Reply:
Yeah I guess we'll have to look at every DB transaction individually to make sure they're either serialized by the wallet's own internal lock or are `reset()` capable if not. I implemented your suggestion about adding the `View()` and `Update()` methods to the DB interface directly. That will make it possible to start migrating/upgrading individual code blocks in the future.